### ISSUE TITLE
[proposal] Make pythongen classes just dataclasses - decouple YAML from class behavior

### ISSUE THREAD

**Issue by sneakers-the-rat** [linkml/linkml#1966]
**Is your feature request related to a problem? Please describe.**
I have so far been sticking to the pydantic side of things and haven't looked too deeply at the python dataclasses models, but in trying to figure out a fix for https://github.com/linkml/linkml-model/pull/189 I noticed that the python models are constructed and used in a way that makes it difficult to decouple their role as a representation of the underlying schema model as YAML/JSON and as runtime classes.

They inherit from `YAMLRoot` - https://github.com/linkml/linkml-runtime/blob/7c311d975422ddf990ae524fe6d969326c1b9261/linkml_runtime/utils/yamlutils.py#L27

which is itself a `JsonObj` from `jsonasobj2`. As far as I can tell the purpose of that inheritance is to be able to use the objects in a such a way that `getitem` and `getattr` are equivalent: eg `item.field == item['field']`. The `YAMLRoot` class enforces the absence of attributes that aren't dataclass fields in its `post_init` method, and throughout linkml they are used as being roughly equivalent to dictionaries - eg. the use of `vars()` to iterate through their contents rather than accessing model fields through something like `fields`, or even better through a method like `slots()` that would be able to iterate specifically through just slots, etc.

There seems to be a decent amount of complexity in the loaders and dumpers in trying to use them as quasi-equivalent to the YAML/JSON representation, and that also prevents us from doing nice things like using the inheritance hierarchy to give ourselves attributes and methods for introspection - eg. to do https://github.com/linkml/linkml/issues/1960 it would be nice to be able to do `myclass.identifier_slot()` to get any identifier slot that a class might have, but that would require us to store additional model metadata in the class, either in a private attribute or in a `field()` annotation. At the point of using the loaders, one no longer has access to the source schema, and the dataclasses don't have enough information to do something like infer which slot is an identifier on their own, so we are precluded from doing some obvious nice-to-have things like making the loaders more flexible in ways that are intuitive from a modeling POV.

**Describe the solution you'd like**
The first part of this is a question - why do the classes inherit from `JsonObj`? I'm probably missing something there, but it seems like it a) makes things brittle, b) requires a bunch of extra convenience functions to use the models, and c) makes the models not behave like one might expect python dataclasses to, which I imagine makes them difficult to use in downstream cases.

The second part is a proposal - if we can keep existing functionality, remove `JsonObj` in the inheritance hierarchy and refactor parent class to be able to load and dump from the JSON/YAML representation without being expected to be a 1:1 in-memory representation of it. Then we can use the parent class to provide convenience methods to pythongen classes (including potentially removing the `post_init` method from each class and replace it with generic validators in the parent class, and putting more class and slot level metadata in the class and slot models themselves).

Pretty broad scope of issue, the purpose here is to try and nail down that scope - what is desirable and possible? what am i missing? 

**How important is this feature?** Select from the options below:
• Medium - can do work without it; but it's important (e.g. to save time or for convenience)

**When will use cases depending on this become relevant?** Select from the options below:
• Mid-term - 2-4 months

**Comment by cmungall** [comment]
I agree with your analysis. (Although I don't think this is the blocker to https://github.com/linkml/linkml-model/pull/189 -- see the discussion I linked to from that ticket).

See similar issues:

- #485 
- #661

`JsonObj` has some clever things in it that can be useful - for example, whenever you make a typo in a schema, your error report will include a line number rather than just a yaml blob, which can be handy.

However, we should ween ourselves off of it and have fewer dependencies, and fewer non-idiomatic pythonisms.

My preferred approach here would be to write a new `DataclassGenerator`, independent of JsonObj, with reasonable parity with `PydanticGenerator`. By reasonable parity I mean don't attempt to recapitulate pydantic's validation capabilities. However, I think they should be introspectable in largely the same way. Groups who wish to use dataclasses without committing to pydantic could then gradually swap out, and the metamodel eventually swapping one way or another.

However I'm open to a top down strategy where we start by swapping out JsonObj as a parent for YAMLRoot

**Comment by ticapix** [comment]
Hello,

I came across this thread thanks to a PR @cmungall pushed on `jsonasobj2`, while I was working on a tiny refactoring there too https://github.com/hsolbrig/jsonasobj2/pull/5

... and I agree with both previous comments: the dependency on `jsonasobj2` is making the loaders/dumpers quite cumbersome and buggy.

See https://github.com/linkml/linkml-runtime/pull/356

I came across another bug yesterday when serializing nested blank nodes.

Would someone take the challenge to remove the `jsonasobj2` dependency:

- are there specific constrains to consider ?
- do we feel confident the existing tests suite will avoid regression ?

**Comment by dalito** [comment]
Maybe the hsolbrig packages used in linkML could be moved under the linkML org to keep them alive? With his retirement, the maintenance of the packages stopped (very well understandable).

**Comment by ticapix** [comment]
As a first step, to avoid creating unwanted side-effects in other parts of the code, what about copying/pasting the single source file `jsonasobj2/_jsonobj.py` under `linkml_runtime/utils/` folder and start refactoring `linkml_runtime/utils/yamlutils.py` bit by bit ?

**Comment by cmungall** [comment]
I think this makes sense.

I have full access to the hsolbrig repos/pypis (or can get them if I don't
already). However, I think moving the code into the linkml-runtime repo
(effectively vendoring) will make it easier to test changes etc, and
eventually remove dependency on them

**Comment by sneakers-the-rat** [comment]
Vendoring sounds good, will also let us see what gets used through code coverage. Can we also add a deprecation warning to the jsonasobj-specific functionality (eg. The convenience functions, accessing via getitem, etc.) When we do that? Want to give people a heads up that we're making that change, bc I think we're agreed that we also don't necessarily want to keep all that behavior (aka I think we want to make the dataclasses behave just like dataclasses) and want to give a heads up we're removing it rather than trying to backport (sideport? idk) it into linkml-runtime forever

**Comment by sierra-moxon** [comment]
from dev call:
- maybe we should just bring the repos themselves into the LinkML organization. (to avoid breaking existing generated code in the wild). 
- or vendor it and then deal with new models going forward from one source, old from old source. 
- maybe we can remove the import (if its unnecessary) as a first step 
```
Imported like this in schemas:

from jsonasobj2 import (
    JsonObj,
    as_dict
)
``` this is one of the bits of code we should be very conservative about changing.

**Comment by cmungall** [comment]
Vendoring could get us into problems.

Let's say I generate my datamodel using pythongen using `vN` of the stack, right now. 

linkml_runtime.utils.yamlutils for version `vN` has this:

```python
from jsonasobj2 import JsonObj, as_json, as_dict, JsonObjTypes, items
import jsonasobj2
...
class YAMLRoot(JsonObj):
```

My python classes generated using `vN` look something like this:

foo.py:
```python
from jsonasobj2 import JsonObj, as_dict
...
from linkml_runtime.utils.yamlutils import YAMLRoot, extended_str, extended_float, extended_int
...

class Foo(YAMLRoot):
    ...
    def __post_init__(...):
        ...
        self.bar = Baz(**as_dict(self.bar))
```

Now, post-vendoring (`vN+1`) we would presumably change yamlutils to be something like:

```python
from linkml_runtime.jsonasobj2 import JsonObj, as_json, as_dict, JsonObjTypes, items
import linkml_runtime.jsonasobj2
...
class YAMLRoot(JsonObj):
```

Now I assume I have a downstream library that uses my Foo model. It's frequent for the runtime to be ahead of generated classes (e.g. when the downstream library does a `poetry update`)

We now have a situation where `vN` dataclasses are mixed with the `vN+1` runtime, so there are two copies of various classes in play

- `linkml_runtime.JsonObj`
- `jsonasobj2.JsonObj`

this means that idioms such as `isinstance(x, JsonObj)` will in some cases silently fail where they previously succeeded

**Comment by sneakers-the-rat** [comment]
Agreed. Lets put a deprecation warning up first

**Comment by ticapix** [comment]
> Vendoring could get us into problems.

because of the circular dependency `linkml-runtime` --depends on--> `linkml-model` --uses--> `linkml.generators.pythongen` --create inputs for--> `linkml-runtime`

Unless we synchronize the merge of 2 PR, one on the repo `linkml-runtime` and one on the repo `linkml`, for the same release, no ?

Here is a proposal to update `linkml-runtime`, which can't be merged before we have one on `linkml` ready to be approved https://github.com/linkml/linkml-runtime/pull/361

I run the tests and measure the coverage. I think we have a decent coverage of JsonObj to start modifying it
![image](https://github.com/user-attachments/assets/5b2fc4ce-20b1-414f-8e03-71a5d40c5098)

Would something like that work ?
